File: cs/Minsk.Tests/CodeAnalysis/Syntax/ParserTests.cs
﻿using Minsk.CodeAnalysis.Syntax;
using Xunit;

namespace Minsk.Tests.CodeAnalysis.Syntax;

public class ParserTests
{
    [Theory]
    [MemberData(nameof(GetBinaryOperatorPairsData))]
    private void BinaryExpressionHonorsPrecedence(SyntaxKind op1, SyntaxKind op2)
    {
        var op1Precedence = SyntaxFacts.GetBinaryOperatorPrecedence(op1);
        var op2Precedence = SyntaxFacts.GetBinaryOperatorPrecedence(op2);
        var op1Text = SyntaxFacts.GetText(op1)!;
        var op2Text = SyntaxFacts.GetText(op2)!;
        var text = $"a {op1Text} b {op2Text} c";
        var expression = ParseExpression(text);

        if (op1Precedence >= op2Precedence)
        {
            using var e = new AssertingEnumerator(expression);

            e.AssertNode(SyntaxKind.BinaryExpression);
            e.AssertNode(SyntaxKind.BinaryExpression);
            e.AssertNode(SyntaxKind.NameExpression);
            e.AssertToken(SyntaxKind.IdentifierToken, "a");
            e.AssertToken(op1, op1Text);
            e.AssertNode(SyntaxKind.NameExpression);
            e.AssertToken(SyntaxKind.IdentifierToken, "b");
            e.AssertToken(op2, op2Text);
            e.AssertNode(SyntaxKind.NameExpression);
            e.AssertToken(SyntaxKind.IdentifierToken, "c");
        }
        else
        {
            using var e = new AssertingEnumerator(expression);

            e.AssertNode(SyntaxKind.BinaryExpression);
            e.AssertNode(SyntaxKind.NameExpression);
            e.AssertToken(SyntaxKind.IdentifierToken, "a");
            e.AssertToken(op1, op1Text);
            e.AssertNode(SyntaxKind.BinaryExpression);
            e.AssertNode(SyntaxKind.NameExpression);
            e.AssertToken(SyntaxKind.IdentifierToken, "b");
            e.AssertToken(op2, op2Text);
            e.AssertNode(SyntaxKind.NameExpression);
            e.AssertToken(SyntaxKind.IdentifierToken, "c");
        }
    }

    [Theory]
    [MemberData(nameof(GetUnaryOperatorPairsData))]
    private void UnaryExpressionHonorsPrecedences(SyntaxKind unaryKind, SyntaxKind binaryKind)
    {
        var unaryPrecedence = SyntaxFacts.GetUnaryOperatorPrecedence(unaryKind);
        var binaryPrecedence = SyntaxFacts.GetBinaryOperatorPrecedence(binaryKind);
        var unaryText = SyntaxFacts.GetText(unaryKind)!;
        var binaryText = SyntaxFacts.GetText(binaryKind)!;
        var text = $"{unaryText} a {binaryText} b";
        var expression = ParseExpression(text);

        if (unaryPrecedence >= binaryPrecedence)
        {
            using var e = new AssertingEnumerator(expression);

            e.AssertNode(SyntaxKind.BinaryExpression);
            e.AssertNode(SyntaxKind.UnaryExpression);
            e.AssertToken(unaryKind, unaryText);
            e.AssertNode(SyntaxKind.NameExpression);
            e.AssertToken(SyntaxKind.IdentifierToken, "a");
            e.AssertToken(binaryKind, binaryText);
            e.AssertNode(SyntaxKind.NameExpression);
            e.AssertToken(SyntaxKind.IdentifierToken, "b");
        }
        else
        {
            using var e = new AssertingEnumerator(expression);
            e.AssertNode(SyntaxKind.UnaryExpression);
            e.AssertToken(unaryKind, unaryText);
            e.AssertNode(SyntaxKind.BinaryExpression);
            e.AssertNode(SyntaxKind.NameExpression);
            e.AssertToken(SyntaxKind.IdentifierToken, "a");
            e.AssertToken(binaryKind, binaryText);
            e.AssertNode(SyntaxKind.NameExpression);
            e.AssertToken(SyntaxKind.IdentifierToken, "b");
        }
    }

    private static IEnumerable<object[]> GetBinaryOperatorPairsData()
    {
        return SyntaxFacts.GetBinaryOperatorKinds()
            .SelectMany(_ => SyntaxFacts.GetBinaryOperatorKinds(),
                (op1, op2) => new object[] { op1, op2 });
    }

    private static IEnumerable<object[]> GetUnaryOperatorPairsData()
    {
        return SyntaxFacts.GetUnaryOperatorKinds()
            .SelectMany(_ => SyntaxFacts.GetBinaryOperatorKinds(),
                (unary, binary) => new object[] { unary, binary });
    }

    private static ExpressionSyntax ParseExpression(string text)
    {
        var statement = SyntaxTree.Parse(text).Root.Statement;
        return Assert.IsType<ExpressionStatementSyntax>(statement).Expression;
    }
}

File: kt/minsk/src/test/kotlin/dev/phytolizer/minsk/analysis/syntax/ParserTests.kt
package dev.phytolizer.minsk.analysis.syntax

import dev.phytolizer.minsk.TestUtils
import io.kotest.core.spec.style.FunSpec
import io.kotest.datatest.withData
import io.kotest.matchers.types.shouldBeTypeOf

class ParserTests : FunSpec({
    context("binary expression honors precedence") {
        withData(nameFn = { "${it[0]} x ${it[1]}" }, binaryOperatorPairs()) {
            val op1Text = SyntaxFacts.getText(it[0])!!
            val op2Text = SyntaxFacts.getText(it[1])!!
            val op1Precedence = SyntaxFacts.binaryOperatorPrecedence(it[0])
            val op2Precedence = SyntaxFacts.binaryOperatorPrecedence(it[1])

            val text = "a $op1Text b $op2Text c"
            val expression = parseExpression(text)

            if (op1Precedence >= op2Precedence) {
                val e = AssertingEnumerator(expression)

                e.assertNode(SyntaxKind.BinaryExpression)
                e.assertNode(SyntaxKind.BinaryExpression)
                e.assertNode(SyntaxKind.NameExpression)
                e.assertToken(SyntaxKind.IdentifierToken, "a")
                e.assertToken(it[0], op1Text)
                e.assertNode(SyntaxKind.NameExpression)
                e.assertToken(SyntaxKind.IdentifierToken, "b")
                e.assertToken(it[1], op2Text)
                e.assertNode(SyntaxKind.NameExpression)
                e.assertToken(SyntaxKind.IdentifierToken, "c")
                e.assertAtEnd()
            } else {
                val e = AssertingEnumerator(expression)

                e.assertNode(SyntaxKind.BinaryExpression)
                e.assertNode(SyntaxKind.NameExpression)
                e.assertToken(SyntaxKind.IdentifierToken, "a")
                e.assertToken(it[0], op1Text)
                e.assertNode(SyntaxKind.BinaryExpression)
                e.assertNode(SyntaxKind.NameExpression)
                e.assertToken(SyntaxKind.IdentifierToken, "b")
                e.assertToken(it[1], op2Text)
                e.assertNode(SyntaxKind.NameExpression)
                e.assertToken(SyntaxKind.IdentifierToken, "c")
                e.assertAtEnd()
            }
        }
    }

    context("unary operator honors precedence") {
        withData(
            nameFn = { "${it[0]} x ${it[1]}" },
            unaryOperatorPairs(),
        ) {
            val unaryKind = it[0]
            val binaryKind = it[1]
            val unaryText = SyntaxFacts.getText(unaryKind)!!
            val binaryText = SyntaxFacts.getText(binaryKind)!!
            val unaryPrecedence = SyntaxFacts.unaryOperatorPrecedence(unaryKind)
            val binaryPrecedence = SyntaxFacts.binaryOperatorPrecedence(binaryKind)
            val text = "$unaryText a $binaryText b"
            val expression = parseExpression(text)

            if (unaryPrecedence >= binaryPrecedence) {
                val e = AssertingEnumerator(expression)

                e.assertNode(SyntaxKind.BinaryExpression)
                e.assertNode(SyntaxKind.UnaryExpression)
                e.assertToken(unaryKind, unaryText)
                e.assertNode(SyntaxKind.NameExpression)
                e.assertToken(SyntaxKind.IdentifierToken, "a")
                e.assertToken(binaryKind, binaryText)
                e.assertNode(SyntaxKind.NameExpression)
                e.assertToken(SyntaxKind.IdentifierToken, "b")
                e.assertAtEnd()
            } else {
                val e = AssertingEnumerator(expression)

                e.assertNode(SyntaxKind.UnaryExpression)
                e.assertToken(unaryKind, unaryText)
                e.assertNode(SyntaxKind.BinaryExpression)
                e.assertNode(SyntaxKind.NameExpression)
                e.assertToken(SyntaxKind.IdentifierToken, "a")
                e.assertToken(binaryKind, binaryText)
                e.assertNode(SyntaxKind.NameExpression)
                e.assertToken(SyntaxKind.IdentifierToken, "b")
                e.assertAtEnd()
            }
        }
    }
}) {
    companion object {
        private fun binaryOperatorPairs() =
            TestUtils.cartesianProduct(SyntaxFacts.binaryOperators(), SyntaxFacts.binaryOperators())

        private fun unaryOperatorPairs() =
            TestUtils.cartesianProduct(SyntaxFacts.unaryOperators(), SyntaxFacts.binaryOperators())

        private fun parseExpression(text: String): ExpressionSyntax =
            SyntaxTree.parse(text).root.statement.shouldBeTypeOf<ExpressionStatementSyntax>().expression
    }
}
